src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,727 @@
+    in
+    match sum_coefficients (get_coeff_vec t texp) with
+    | exception _ -> None
+    | None -> None
+    | Some (var, var_coeff, offset) ->
+      if Option.is_none var then Some (None, offset)
+      else if Z.(var_coeff = one) then Some (var, offset)
+      else None
+
+
+  let get_coeff t texp = timing_wrap "coeff_vec" (get_coeff t) texp
+
+  let abstract_exists var t = match t.d with
+    | Some d -> {t with d = Some (EArray.reduce_col d (Environment.dim_of_var t.env var))}
+    | None -> t (* there are no  variables in the current environment *)
+
+  let assign_const t var const = match t.d with
+    | None -> t
+    | Some t_d -> let d = EArray.copy t_d in d.(var) <- (None, const);  {d = Some d; env = t.env}
+
+  let subtract_const_from_var t var const =
+    match t.d with
+    | None -> t
+    | Some t_d -> let d = EArray.copy t_d in
+      let subtract_const_from_var_for_single_equality index element =
+        let (eq_var_opt, off2) = d.(index) in
+        if index = var then
+          match eq_var_opt with
+          | None -> d.(index) <- (None, Z.(off2 + const))
+          | Some eq_var -> begin if eq_var <> index then d.(index) <- (eq_var_opt, Z.(off2 + const)) end
+        else
+          begin match eq_var_opt with
+            | Some eq_var ->
+              if eq_var = var then d.(index) <- (eq_var_opt, Z.(off2 - const))
+            | None -> ()
+          end
+      in
+      EArray.iteri (subtract_const_from_var_for_single_equality) d; {d = Some d; env = t.env}
+end
+
+
+(** TODO: overflow checking *)
+module ExpressionBounds: (SharedFunctions.ExtendedConvBounds with type t = VarManagement.t) =
+struct
+  include VarManagement
+
+  let bound_texpr t texpr = let texpr = Texpr1.to_expr texpr in
+    match get_coeff t texpr with
+    | Some (None, offset) -> Some offset, Some offset
+    | _ -> None, None
+
+
+  let bound_texpr d texpr1 =
+    let res = bound_texpr d texpr1 in
+    match res with
+    | Some min, Some max ->  if M.tracing then M.tracel "bounds" "min: %s max: %s" (IntOps.BigIntOps.to_string min) (IntOps.BigIntOps.to_string max); res
+    | _ -> res
+
+  let bound_texpr d texpr1 = timing_wrap "bounds calculation" (bound_texpr d) texpr1
+end
+
+module D =
+struct
+  include Printable.Std
+  include ConvenienceOps (Mpqf)
+  include VarManagement
+
+  module Bounds = ExpressionBounds
+
+  module Convert = SharedFunctions.Convert (V) (Bounds) (struct let allow_global = true end) (SharedFunctions.Tracked)
+
+  type var = V.t
+
+  let name () = "lin2vareq"
+
+  let to_yojson _ = failwith "ToDo Implement in future"
+
+  let is_bot t = equal t (bot ())
+  let is_bot_env t = t.d = None
+
+  (*this shows "top" for a specific environment to enable the calculations. It is the top_of of all equalities*)
+  let top_of env = {d = Some (EArray.make_empty_array (Environment.size env)); env = env}
+
+  (*Should never be called but implemented for completeness *)
```

Comment:
In which sense "should" this never be called? It is not exepected to be called? Or it would be a problem if the other code would have to deal with the `top ()` element?

Reply:
It is not expected to be called. In affeq this function simply throws an exception. We decided to implement something meaningful instead of the exception, even though we don't expect it to be called.